### ISSUE TITLE
[FEATURE REQUEST] Model Fine-Tuning

### ISSUE THREAD

**Issue by suuuehgi** [Akylas/OSS-DocumentScanner#564]
**Is your feature request related to a problem? Please describe.**
Better document recognition. The auto-cropping of isolated documents work well, but the cropping of e.g. single pages of a book need some help.

**Describe the solution you'd like**
Implement a fine-tuning loop based on the usage of the app.

Thousands of people manually adjust / improve the cropping region on their phones.
That could be used as fine-tune training data for further improvement.

Probably the easiest way is to add a toggle option for data donations in the settings and collect the data.
Better but I don't know if possible: Do it on the users phone and send back some anonymous data without sensitive informations.

**Describe alternatives you've considered**
---

**Additional context**

![Image](https://github.com/user-attachments/assets/61cf76e6-3dc4-41cb-9dea-65fad71cf2d3)

P. S. Thank you for the already amazing work!
P. P. S. Or are there already existing cropping data sets for book pages?

**Comment by farfromrefug** [comment]
@suuuehgi i wont add any data collection to the app.
I understand your issue but book pages detection is a very tricky problem. Yes we could use machine learning to crop the image.
I am not against the idea at all. But the tricky part is training a model with enough data to fit "all" users requirements.
If you can find such a model i would be happy to test it

**Comment by suuuehgi** [comment]
I would love to look into this, but I don't know whether I realistically find time in the near future. At the moment I'm completely booked and additionally I would have to dive into the source code, architecture, etc. It's quite a few years ago that I trained and fine tuned models and I never really dealt with Android app programming.

**Comment by egdels** [comment]
> [@suuuehgi](https://github.com/suuuehgi) i wont add any data collection to the app. I understand your issue but book pages detection is a very tricky problem. Yes we could use machine learning to crop the image. I am not against the idea at all. But the tricky part is training a model with enough data to fit "all" users requirements. If you can find such a model i would be happy to test it

I recently trained such a model for MakeACopy. It would be an honor if it could serve as a foundation for a model used in OSS DocumentScanner.

https://github.com/egdels/makeacopy/tree/main/training

**Comment by farfromrefug** [comment]
@egdels that s amazing ! Could you guide me a bit in testing this? First i would like to test it on my desktop using python. Do you also have a set of test images? Would love to see how it handles all kind of images, like books for example.
When scanning a book where we see the 2 pages. Will it return the "corners" for each pages (maybe we can even scan both pages at the same time).

**Comment by egdels** [comment]
@farfromrefug  

Great, happy to guide you 🙂

DocQuadNet-256 predicts **one quadrilateral (4 corners) per image** — the most likely document region.

For an open book (two visible pages), it will typically:

- return a quad around **both pages together**, or  
- sometimes around **one page**, depending on contrast, edges and postprocessing.

It does **not** return two separate page quads out of the box.

Supporting “scan both pages separately” would require an additional step, for example:

- gutter detection + splitting into two regions of interest (ROIs), or  
- a dedicated two-page mode/model.

You can test the ONNX model directly on your desktop using the existing evaluation pipeline.

I recommend testing with your own representative photos (including open books).

You can quickly label a small test set using the included tool:

```
python3 training/scripts/labeler.py \
  --img_dir training/data/my_test_images \
  --out training/labels/my_test_labels.jsonl
```

Convert to Trainable Format

```
python3 training/scripts/convert_to_docquad.py \
  --in_images training/data/my_test_images \
  --in_jsonl training/labels/my_test_labels.jsonl \
  --out_dir training/data/my_test_converted

python3 training/scripts/make_trainable_from_converted.py \
  --in_images training/data/my_test_converted/images \
  --in_labels training/data/my_test_converted/labels \
  --out_dir training/data/my_test_trainable
```

Run Evaluation

```
python3 training/scripts/evaluate_docquad_models.py \
  --test_dir training/data/my_test_trainable \
  --model docquad=app/src/main/assets/docquad/docquadnet256_trained_opset17.onnx \
  --out /tmp/docquad_eval.json \
  --mode product
```

This will:

- run inference
- apply the same postprocessing logic used in the app
- compute corner MAE / IoU
- report failure reasons (self-intersecting, non-convex, degenerate-area, etc.)

You can start with just 10–20 images to quickly see how it behaves before building a larger test set.

**Comment by egdels** [comment]
Small addition:

You can download the ONNX model directly here: https://github.com/egdels/makeacopy/raw/c9330fd313f24caf78e7fb2ee85c690a348dc0c0/app/src/main/assets/docquad/docquadnet256_trained_opset17.onnx

For desktop testing, please use this .onnx model (the .ort version is optimized for Android deployment).

**Comment by egdels** [comment]
@farfromrefug 

I created a small prototype fork integrating DocQuadNet-256 into OSS Document Scanner:

https://github.com/egdels/OSS-DocumentScanner

Important:
This is purely experimental and not production-ready.
It is meant to explore whether ML-based corner detection could be a viable complement to the existing approach.

In many normal cases, the current implementation already performs very well.
However, I observed some extreme or low-contrast cases where the ML model behaves more robustly.

I do not claim it is universally better - it is just a prototype for evaluation and discussion.

Thanks for maintaining this project. Keep up the good work!

**Comment by farfromrefug** [comment]
@egdels ok that s awesome. Thank you i will test it. About the gutter detection, this is something i am really interested about and asked by many users. Do you think it could be possible to teach the model for that?

**Comment by egdels** [comment]
We could keep DocQuadNet as a single-quad detector and implement gutter splitting purely in OpenCV on the rectified image. That avoids multi-quad training and keeps the ML part simple.

**Comment by farfromrefug** [comment]
@egdels that s a possibility. I tried gutter detection in opencv but it was not working well:
* i was detecting it on "source" image so it was peeking book "borders" and other features
*  when the page have text columns it often picks that up as a gutter

Here are 2 images i am struggling with.
Did not have time to test your code yet. Will do!

**Comment by farfromrefug** [comment]
@egdels i looked at the readme in your fork. Great job in the integration!
onnxruntime seems really big to me, i dont really like that. 
Have you ever looked at https://github.com/Tencent/ncnn ? 
It seems very lightweight while still very good especially for vision.
Do you think we could use that instead with your model?

**Comment by egdels** [comment]
@farfromrefug

I looked into ncnn a bit while experimenting with the integration.

I converted the model with PNNX and compared it directly against ONNX Runtime on a small SmartDoc validation subset (75 images). Unfortunately, the results were not equivalent.

With ONNX Runtime the model sits around ~25px mean corner error.  
With ncnn (via PNNX) it was around ~299px.

That’s not a small numerical drift - even the raw tensor outputs already differ quite a lot before any postprocessing. I tried default PNNX settings, disabling optimizations, FP16 vs FP32, but I couldn’t get it to match the ONNX reference for this architecture (MobileNetV3 + FPN + several Resize ops).

So at least in its current form, I wouldn’t feel comfortable switching this model to ncnn.

Regarding size: I completely understand the concern about onnxruntime. I’m still figuring out the cleanest mobile packaging approach, but ORT can actually be built in a very minimal configuration.

I haven’t finalized that part yet, but I can try to prepare a minimal build script if we decide the model is worth integrating.

Maybe the best next step is to first test the ONNX model on desktop or whin proto with your book images and see whether it actually improves those cases. If it doesn’t bring clear value, there’s no point optimizing runtime size 🙂

If it does look promising, I’m happy to work through the mobile side properly.

**Comment by farfromrefug** [comment]
@egdels thanks for the detailed explanation. and yes the first step is to test it locally.
need to find the time.
thank you

**Comment by egdels** [comment]
No rush. I’ll keep working on it. I think I can reduce the runtime to about 1.5 MB per ABI. If I manage to make that reproducible, my own project will benefit as well. I’ve started experimenting with book gutter detection using OpenCV, but the results don’t look very good yet.

**Comment by farfromrefug** [comment]
@egdels awesome! About the model I think I would store it on github releases and have the app download it when the user decides to use it.

Yes gutter detection is pretty tricky, and I can't really find references on the web about working solutions

**Comment by suuuehgi** [comment]
I stumbled upon some other models that are below 15MB: 
https://docsaid.org/en/docs/docaligner/benchmark/

**Comment by egdels** [comment]
My project used the DocAligner model up to version 3.x.y.
The problem was that I couldn’t fine-tune it. In addition, the ONNX model could not be converted to ORT format, which made the runtime larger again.

In terms of results, I was satisfied with DocAligner. However, the model is not reproducible for me.

At https://github.com/egdels/makeacopy/tree/main/training

I describe how to train your own model. That way, you also have the checkpoints available, so you can continue training later — for fine-tuning.

**Comment by egdels** [comment]
```
Searching for final AAR ...
Copied: ./makeacopy/app/libs/onnxruntime-android-v1.24.1.aar  (size: 3,8M)

DONE.
unzip -lv app/libs/onnxruntime-android-v1.24.1.aar \
| awk '/libonnxruntime/ { printf "%-45s  uncompressed=%6.2f MB  compressed=%6.2f MB\n", $8, $1/1024/1024, $3/1024/1024 }'
jni/arm64-v8a/libonnxruntime.so                uncompressed=  2,24 MB  compressed=  0,92 MB
jni/arm64-v8a/libonnxruntime4j_jni.so          uncompressed=  0,11 MB  compressed=  0,03 MB
jni/armeabi-v7a/libonnxruntime.so              uncompressed=  1,41 MB  compressed=  0,71 MB
jni/armeabi-v7a/libonnxruntime4j_jni.so        uncompressed=  0,08 MB  compressed=  0,03 MB
jni/x86/libonnxruntime.so                      uncompressed=  2,04 MB  compressed=  0,86 MB
jni/x86/libonnxruntime4j_jni.so                uncompressed=  0,09 MB  compressed=  0,03 MB
jni/x86_64/libonnxruntime.so                   uncompressed=  2,41 MB  compressed=  0,95 MB
jni/x86_64/libonnxruntime4j_jni.so             uncompressed=  0,09 MB  compressed=  0,03 MB

```
I will provide the build script in the fork.

**Comment by suuuehgi** [comment]
> Yes gutter detection is pretty tricky, and I can't really find references on the web about working solutions

I just stumbled upon [this one](https://computeroptics.ru/eng/KO/Annot/KO48-4/480409e.html). It's actually about folding, but my gut feeling says a gutter is not too different.

**Comment by farfromrefug** [comment]
@egdels if i want to simply test your onnx model on desktop with input images, do you have a simple python script to do that? 
ideally i want to open an image with opencv, run your model, draw the detected corners.

**Comment by egdels** [comment]
I’ve updated the fork and will add a simple script to test images locally.

**Comment by egdels** [comment]
https://raw.githubusercontent.com/egdels/OSS-DocumentScanner/refs/heads/main/docquad-integration/README.md

https://raw.githubusercontent.com/egdels/OSS-DocumentScanner/refs/heads/main/docquad-integration/scripts/visualize_corner_detection.py

**Comment by farfromrefug** [comment]
@egdels you are awesome and fast ! will try

**Comment by farfromrefug** [comment]
@egdels i tested your model on some sample images. If the "document" is following common rules it seems to be very good. 
However when there are some edge cases it does not work so well.
Here is part of the test images i use.
Maybe the model can be improved to handles such cases? My opencv implementation handles those quite "well enough"

[test_images.zip](https://github.com/user-attachments/files/25459579/test_images.zip)

**Comment by egdels** [comment]
I’m trying to have fun with it.
Just a quick note: for MakeACopy, the model is currently sufficient. For OSS Document Scanner, it could be a good foundation if it’s further fine-tuned. Everything needed for that already exists — except individual training images.

![Image](https://github.com/user-attachments/assets/110950ce-5e80-4ab3-9ab2-c0f747f70ed8)

![Image](https://github.com/user-attachments/assets/535f7228-eb0a-410f-a5b1-85c6eec2522f)

**Comment by egdels** [comment]
If no corners are present - like in tax.jpeg - the model cannot detect any. That’s a case for postprocessing.

In MakeACopy, I use the model first, and if it fails, I fall back to OpenCV.

**Comment by farfromrefug** [comment]
@egdels I agree it works great for your use case. For my app I need a more complete solution handling most cases. I think the only remaining one really is the opened book.

**Comment by egdels** [comment]
![Image](https://github.com/user-attachments/assets/58ca456e-75ed-492a-bb86-b38ded94b455)

> [@egdels](https://github.com/egdels) that s a possibility. I tried gutter detection in opencv but it was not working well:
> 
>     * i was detecting it on "source" image so it was peeking book "borders" and other features
> 
>     * when the page have text columns it often picks that up as a gutter
> 
> 
> Here are 2 images i am struggling with. Did not have time to test your code yet. Will do!

Can you provide the test images as well? I’m currently trying to build a prototype in Python using the hybrid approach.

![Image](https://github.com/user-attachments/assets/3472796c-0af9-4820-b8d2-71fd005ad6e9)

**Comment by farfromrefug** [comment]
@egdels I shared some test images in a previous comment.
your test looks promising
